pkg/reslife/lifecycle.go

```diff
@@ -0,0 +1,256 @@
+package reslife
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/justtrackio/gosoline/pkg/appctx"
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/clock"
+	"github.com/justtrackio/gosoline/pkg/funk"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+type (
+	LifeCycleer interface {
+		GetId() string
+	}
+	Creator interface {
+		Create(ctx context.Context) error
+	}
+	Initializer interface {
+		Init(ctx context.Context) error
+	}
+	Registerer interface {
+		Register(ctx context.Context) (string, any, error)
+	}
+	Purger interface {
+		Purge(ctx context.Context) error
+	}
+	LifeCycleerFactory func(ctx context.Context, config cfg.Config, logger log.Logger) (LifeCycleer, error)
+	managerCtxKey      struct{}
+)
+
+type Settings struct {
+	Create struct {
+		Enabled bool `cfg:"enabled" default:"false"`
+	} `cfg:"create"`
+	Init struct {
+		Enabled bool `cfg:"enabled" default:"true"`
+	} `cfg:"init"`
+	Register struct {
+		Enabled bool `cfg:"enabled" default:"true"`
+	} `cfg:"register"`
+	Purge struct {
+		Enabled bool `cfg:"enabled" default:"false"`
+	} `cfg:"purge"`
+}
+
+type LifeCycleManager struct {
+	logger     log.Logger
+	clock      clock.Clock
+	settings   *Settings
+	refresher  func(fac LifeCycleerFactory) (LifeCycleer, error)
+	resources  []LifeCycleer
+	created    []string
+	registered []string
+	purged     []string
+}
+
+func ProvideLifeCycleManager(ctx context.Context, config cfg.Config, logger log.Logger) (*LifeCycleManager, error) {
+	return appctx.Provide(ctx, managerCtxKey{}, func() (*LifeCycleManager, error) {
+		logger = logger.WithChannel("lifecycle-manager")
+
+		settings := &Settings{}
+		config.UnmarshalKey("resource_lifecycles", settings)
+
+		return &LifeCycleManager{
+			logger:   logger,
+			clock:    clock.Provider,
+			settings: settings,
+			refresher: func(fac LifeCycleerFactory) (LifeCycleer, error) {
+				return fac(ctx, config, logger)
+			},
+			resources:  nil,
+			created:    []string{},
+			registered: []string{},
+			purged:     []string{},
+		}, nil
+	})
+}
+
+func (m *LifeCycleManager) refreshResources(ctx context.Context) error {
+	var err error
+	var container *[]LifeCycleerFactory
+	var resource LifeCycleer
+
+	if container, err = provideContainer(ctx); err != nil {
+		return fmt.Errorf("could not get lifecyleers: %w", err)
+	}
+
+	for _, factory := range *container {
+		if resource, err = m.refresher(factory); err != nil {
+			return fmt.Errorf("could not build lifecycleer: %w", err)
+		}
+
+		m.resources = append(m.resources, resource)
```

Comment:
So refresh extends our resources on every iteration? That feels... wrong. But not sure if it is actually wrong

Reply:
It's not beautiful but as we don't know which of those life cycles get called first, this is the safe way to do it.. maybe we can make it nicer in the next iteration..